service/resource/endpoint/desired.go

```diff
@@ -1,7 +1,41 @@
 package endpoint
 
-import "context"
+import (
+	"context"
+
+	"github.com/giantswarm/microerror"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
 
 func (r *Resource) GetDesiredState(ctx context.Context, obj interface{}) (interface{}, error) {
-	return nil, nil
+	endpoint, err := toEndpoint(obj)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	desiredEndpoints := []Endpoint{
+		*endpoint,
+	}
+
+	k8sEndpoints, err := r.k8sClient.CoreV1().Endpoints(endpoint.ServiceNamespace).Get(endpoint.ServiceName, metav1.GetOptions{})
+	if errors.IsNotFound(err) {
```

Comment:
The k8s Endpoint doesn't exist yet, so we only want to add our new Entry.

Reply:
I don't follow here. 

- The custom object is always a pod because the informer is set up accordingly. 
- Using the current state (which I presume you want to do based on this code) to make up the desired state should never be done because the concept of current and desired state is fundamentally different. 
- I feel like we are moving into the wrong direction. I think `GetCurrentState` should fetch the real endpoint with the real IP information. `GetDesiredState` should do what `GetCurrentState` is doing and construct the patch state using the IP that is potentially missing in the real endpoint by looking up the pod annotations. We would then compare the IPs in the other state creation methods and decide which IPs to forward to the processing methods. I apparently lead you into the wrong direction a bit.

Reply:
Mhm I see, I'll restructure this and ping again afterwards.

Reply:
Okay so I hope this is clearer now in my head and makes more sense:
 - `GetCurrentState` fetches the pod annotations first, but only to get information about the related service. It therefore also cancels the reconciliation if the annotations do not exist, as it's first in line.
- `GetCurrentState` compiles a List of all unique IPs attached to the service which describes the current state of the service.
- `GetDesiredState` also fetches the pod annotations to get the desired IP and service information. It does not worry about cancelling the context as it's second in line.
- `GetDesiredState` then creates a List with a single IP to describe the patch set of a missing/to be removed IP.

So the next steps as I see them would be:
- `GetCreateState` creates a List of a unique concatenation of  `currentState` and `desiredState` (return `nil` if length of new list is equal to `currentState`).
- `ProcessCreateState` makes sure the `createState` is reflected in the k8s Endpoint.
- `DeleteState` similar to create, but determine if `currentState` and `desiredState` share an entry.


Let me know if there is any obvious flaw in my thinking. I am not very happy with both the `GetCurrentState` and ` GetDesiredState` using the `Pod` for information, but I didn't see a way to prevent this "nicely".